### PR TITLE
Some build simplification, implement MQCSP auth

### PR DESCRIPTION
**Pull request by rianmcguire**
I recommend reviewing commit-by-commit.

Some relevant docs:
* https://www.ibm.com/docs/en/ibm-mq/9.2?topic=mqcno-securityparmsptr-pmqcsp
* https://www.ibm.com/support/pages/how-specify-userid-and-password-when-connecting-ibm-mq

### CODE REVIEW

File: ext/wmq_queue_manager.c [hunk 1]
```diff
@@ -243,8 +246,22 @@ VALUE QueueManager_initialize(VALUE self, VALUE hash)
...
-        WMQ_HASH2MQCHARS(hash,user_identifier,             pmqcd->UserIdentifier)
-        WMQ_HASH2MQCHARS(hash,password,                    pmqcd->Password)
+
+        /* MQCSP connection authentication with user ID and password */
+        VALUE user_identifier = rb_hash_aref(hash, ID2SYM(ID_user_identifier));
+        VALUE password = rb_hash_aref(hash, ID2SYM(ID_password));
+        if (!NIL_P(user_identifier))
+        {
+            pqm->security_parameters.AuthenticationType = MQCSP_AUTH_USER_ID_AND_PWD;
+            pqm->security_parameters.CSPUserIdPtr = StringValuePtr(user_identifier);
```
Comment (tomdalling) [inline ext/wmq_queue_manager.c]: Do these strings need to be copied and freed, the same way `long_remote_user_id_ptr` is (in this file)? I also noticed that `long_remote_user_id_ptr` gets null terminated, but I'm not sure if that's important.

**Review by tomdalling — APPROVED** [review]
👍 

The amount of C in the gem is a little bit worrying from a maintainability perspective. Any feeling for how many more modifications we're going to need before it's usable?

File: ext/wmq_queue_manager.c (hunk 1 above)
Comment (rianmcguire) [reply ext/wmq_queue_manager.c]: Good point - the GC could run between `initialize` and `connect` and remove those Ruby Strings from the heap.

Definitely don't need to null terminate them through - the MQ library takes a length.

**Comment by rianmcguire** [comment]
> The amount of C in the gem is a little bit worrying from a maintainability perspective. Any feeling for how many more modifications we're going to need before it's usable?

I agree. The only other thing I'm aware of right now is wrapping the blocking `MQGET` call in `rb_thread_call_without_gvl`, so we can use it in a multi-threaded Ruby process. But who knows what else we'll find once we start using it in anger.

**Pull request merged**